File: 2024/python/day01/part2.py
# part2.py
import sys
from collections import Counter


def main():
    """
    Main function to read the input file, process the entries, and calculate
    the similarity score, with a metric based on how often numbers in the left
    list show up in the right list.
    """
    filename = sys.argv[1]
    try:
        with open(filename) as file:
            similarity = 0
            left_list = []
            right_list = []

            for line in file:
                entries = line.strip().split() # strips newlines and spilts by spaces
                if entries:
                    left_list.append(entries[0])
                    right_list.append(entries[1])

        right_counter = Counter(right_list)
        similarity = sum(int(id) * right_counter[id] for id in left_list)

        print(f"Total similarity score: {similarity}")
    except FileNotFoundError:
        print(f"Error: The file {filename} was not found.")

if __name__ == "__main__":
    main()

File: 2024/cpp/day01/part2.cpp
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <vector> 
#include <algorithm> // sorting algo
#include <cmath> // using abs()
#include <sstream>
using namespace std;

int main(int argc, char* argv[]){
    // Open input file for reading
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <input_file>" << endl;
        return 1;
    }
    ifstream filein(argv[1]);
    if (!filein) {
        cerr << "Error opening input file" << endl;
        return 1;
    }
    
    // initialize variables for left and right list, and line for file.
    string line;
    vector<int> left_list;
    vector<int> right_list;
    
    // Pre-allocate memory to avoid reallocations
    left_list.reserve(1000);
    right_list.reserve(1000);
    
    while (getline(filein, line)) {
        istringstream iss(line); 
        int left, right;
        if (iss >> left >> right) {
            left_list.push_back(left);
            right_list.push_back(right);
        }
    }

    // sort both lists
    std::sort(left_list.begin(), left_list.end());
    std::sort(right_list.begin(), right_list.end());

    // calculate pairs and similarity score
    int similarity_score = 0;
    for (int i = 0; i < left_list.size(); i++) {
        int count_appearances = count(right_list.begin(), right_list.end(), left_list[i]);
        similarity_score += count_appearances * left_list[i];
    }

    cout << "Total Simalrity: " << similarity_score << endl;
    filein.close(); // close the file
    return 0;
}
